core/finality-grandpa/src/communication.rs

```diff
@@ -0,0 +1,290 @@
+
+fn localized_payload<E: Encode>(round: u64, set_id: u64, message: &E) -> Vec<u8> {
+	let mut v = message.encode();
+
+	round.using_encoded(|s| v.extend(s));
+	set_id.using_encoded(|s| v.extend(s));
+
+	v
+}
+
+// check a message.
+fn check_message_sig<Block: BlockT>(
+	message: &Message<Block>,
+	id: &AuthorityId,
+	signature: &ed25519::Signature,
+	round: u64,
+	set_id: u64,
+) -> Result<(), ()> {
+	let as_public = ::ed25519::Public::from_raw(id.0);
+	let encoded_raw = localized_payload(round, set_id, message);
+	if ::ed25519::verify_strong(signature, &encoded_raw, as_public) {
+		Ok(())
+	} else {
+		debug!(target: "afg", "Bad signature on message from {:?}", id);
+		Err(())
+	}
+}
+
+/// converts a message stream into a stream of signed messages.
+/// the output stream checks signatures also.
+pub(crate) fn checked_message_stream<Block: BlockT, S>(
+	round: u64,
+	set_id: u64,
+	inner: S,
+	voters: Arc<HashMap<AuthorityId, u64>>,
+)
+	-> impl Stream<Item=SignedMessage<Block>,Error=Error> where
+	S: Stream<Item=Vec<u8>,Error=()>
+{
+	inner
+		.filter_map(|raw| {
+			let decoded = SignedMessage::<Block>::decode(&mut &raw[..]);
+			if decoded.is_none() {
+				debug!(target: "afg", "Skipping malformed message {:?}", raw);
+			}
+			decoded
+		})
+		.and_then(move |msg| {
+			// check signature.
+			if !voters.contains_key(&msg.id) {
+				debug!(target: "afg", "Skipping message from unknown voter {}", msg.id);
+				return Ok(None);
+			}
+
+			// we ignore messages where the signature doesn't check out.
+			let res = check_message_sig::<Block>(
+				&msg.message,
+				&msg.id,
+				&msg.signature,
+				round,
+				set_id
+			);
+			Ok(res.map(move |()| msg).ok())
+		})
+		.filter_map(|x| x)
+		.map_err(|()| Error::Network(format!("Failed to receive message on unbounded stream")))
+}
+
+struct OutgoingMessages<Block: BlockT, N: Network> {
+	round: u64,
+	set_id: u64,
+	locals: Option<(Arc<ed25519::Pair>, AuthorityId)>,
+	sender: mpsc::UnboundedSender<SignedMessage<Block>>,
+	network: N,
+}
+
+impl<Block: BlockT, N: Network> Sink for OutgoingMessages<Block, N> {
+	type SinkItem = Message<Block>;
+	type SinkError = Error;
+
+	fn start_send(&mut self, msg: Message<Block>) -> StartSend<Message<Block>, Error> {
+		// when locals exist, sign messages on import
+		if let Some((ref pair, local_id)) = self.locals {
+			let encoded = localized_payload(self.round, self.set_id, &msg);
+			let signature = pair.sign(&encoded[..]);
+			let signed = SignedMessage::<Block> {
+				message: msg,
+				signature,
+				id: local_id,
+			};
+
+			// forward to network and to inner sender.
+			self.network.send_message(self.round, self.set_id, signed.encode());
+			let _ = self.sender.unbounded_send(signed);
+		}
+
+		Ok(AsyncSink::Ready)
+	}
+
+	fn poll_complete(&mut self) -> Poll<(), Error> { Ok(Async::Ready(())) }
+
+	fn close(&mut self) -> Poll<(), Error> {
+		// ignore errors since we allow this inner sender to be closed already.
+		match self.sender.close() {
```

Comment:
Could be written shorter as:
`self.sender.close().or_else(|_| Ok(Async::Ready()))`